src/Core/Services/Implementations/StripePaymentService.cs

```diff
@@ -956,7 +956,25 @@ public async Task<string> PayInvoiceAfterSubscriptionChangeAsync(ISubscriber sub
                         return paymentIntentClientSecret;
                     }
 
-                    await invoiceService.VoidInvoiceAsync(invoice.Id, new InvoiceVoidOptions());
+                    invoice = await invoiceService.VoidInvoiceAsync(invoice.Id, new InvoiceVoidOptions());
+
+                    // HACK: Workaround for customer balance credit
+                    if (invoice.StartingBalance < 0)
+                    {
+                        // Customer had a balance applied to this invoice. Since we can't fully trust Stripe to
+                        //  credit it back to the customer (even though their docs claim they will), we need to
+                        //  check that balance against the current customer balance and determine if it needs to be re-applied
+                        customer = await customerService.GetAsync(subscriber.GatewayCustomerId, customerOptions);
+
+                        // Assumption: Customer balance should now be $0, otherwise payment would not have failed.
+                        if (customer.Balance == 0)
```

Comment:
What if they had more credit than this invoice was for? Wouldn't it still be > 0?

Reply:
the payment would not have failed if they had a higher credit balance than the invoice amount (already verified this by accident as I misjudged the proration amount when I first tested this)